Prevent bad loading

floffy-f opened the pull request:
# Prevent user from misusing the loading functions

## Strategy :

We let the user go back to the home screen, with an error message.

## Messages :

* When the user tries to *load a single file*, it will get the following message :
![single_file_load](https://user-images.githubusercontent.com/85450305/121696103-04f16480-cacc-11eb-8667-2be46912aa43.png)

* When the user tries to load a *bad file format*, if rust is able to send the appropriate log (some next feature could be to verify that every bad image format is well detected by a rust error), while the file is loading the following logs will appear :
![bad_format](https://user-images.githubusercontent.com/85450305/121696386-4f72e100-cacc-11eb-8ae0-64bc16d52532.png)

* Any other error logs appearing during the loading wil get in this logs section of the loading screen.

## Navigation options

* A button lets the user go back to home screen to try other files. Another one lets it clear the logs definitely too.

:warning: When error logs appear during the loading, if they are not explicitely cleared during that phase, they remain visible on the logs screen of the app after. So thanks to the functionality of the visual marker of the logs in the `headerBar` ( #10 ) we have a red alert visible for the user.
I can't tell if it is relevant or not, maybe the error logs should be removed after coming for the first time in the second section of the app, I prefer letting the app owner choose.

mpizenberg commented:
Thanks @floffy-f for tackling this issue!

The strategy of going back to the home screen where they can re-upload images is the good strategy since it's the only thing that will enable user to make progress. I'd even argue that we can do it without adding a button, since it's one click less and the only thing that a user can do.

Regarding the error reporting, it's maybe more adequate to not split it in two locations depending on the error and instead always use the logging system for this. As such, there is no need for a dedicated `logs` field in the `Logging` state and we can use directly the one of the `Model`. If you want to generate an error log if there is only one image, you can directly do something like

```elm
{ model | seenLogs = thisErrorLog :: model.seenLogs }
```

Where `thisErrorLog` contains the message you want. We can assume when the images are successfully loaded that error logs where seen already.

mpizenberg commented:
By the way, you'll also have to rebase the relevant code in this PR on the current `main`. It will probably be easier to just reset it and copy paste the relevant code.

floffy-f commented:
## Small upgrades
Commit 4e2533c329725fd62bac6088d4cfb9abbfb10524

As suggested, when an error occurs during the loading time, the user is automaticaly sent back to home screen.
The logs are then shown :
![bad_format](https://user-images.githubusercontent.com/85450305/122190368-0aff9080-ce92-11eb-8603-38d44bebd9fc.png)

If the user then retries to load new images, logs are cleared for now to avoid looping in `Home` state.
This will be fixed when I will work on merging if I can.

## No logs in `Loading` state's arguments

The logs are only in the model.
This fixes a little previous issue for which the clean logs button would not reset the view when clicked in `Loading` state.

floffy-f commented:
## Quick precision :

I would suggest to keep the `SingleFile` state as a non-log message, because most of the time if the user loads a single file it is due to a simple missclick, I guess I would prefer not keeping such a message in the logs to keep enough visibility for some more algorithm-oriented errors.

mpizenberg commented:
> ## Quick precision :
> 
> I would suggest to keep the `SingleFile` state as a non-log message, because most of the time if the user loads a single file it is due to a simple missclick, I guess I would prefer not keeping such a message in the logs to keep enough visibility for some more algorithm-oriented errors.

Alright the argument makes sense :+1:

mpizenberg commented:
But then you have to duplicate the logic that is used to handle the home page in the single image page. Since otherwise, the user will have to click again a button to go back to home.

floffy-f committed:
[Prevent bad loading]
Reformat code a little bit + rebase main by hand

mpizenberg committed:
Reload the page when clicking on the return home button

mpizenberg committed:
Fix forgotten log report on wasm image decoding error

mpizenberg committed:
Use logs when only 1 image was uploaded

mpizenberg committed:
Do not show logs while loading

If an error logs is detected,
the application will go back to the home page anyway.

mpizenberg committed:
Small LogState cleanup

mpizenberg committed:
Cleanup some code with logs

mpizenberg committed:
Actually make dedicated loading error page

mpizenberg committed:
Re-add woops loading button

mpizenberg committed:
Fix redirection to LoadingError
